crates/engine/src/prepare/trusted_documents.rs

```diff
@@ -10,14 +10,36 @@ use futures::{future::BoxFuture, FutureExt};
 use grafbase_telemetry::grafbase_client::X_GRAFBASE_CLIENT_NAME;
 use runtime::trusted_documents_client::TrustedDocumentsError;
 use std::borrow::Cow;
+use tracing::Instrument;
 
 use super::PrepareContext;
 
 type DocumentFuture<'a> = BoxFuture<'a, Result<Cow<'a, str>, GraphqlError>>;
 
 pub(crate) struct OperationDocument<'a> {
     pub cache_key: Key<'a>,
-    pub load_fut: DocumentFuture<'a>,
+    pub document_or_future: DocumentOrFuture<'a>,
+}
+
+impl<'a> DocumentOrFuture<'a> {
+    pub(super) async fn into_document(self) -> Result<Cow<'a, str>, GraphqlError> {
+        match self {
+            DocumentOrFuture::Document(cow) => Ok(cow),
+            DocumentOrFuture::Future(future) => {
+                let span = tracing::info_span!("load trusted document");
+                future.instrument(span).await
+            }
+        }
+    }
+}
+
+pub(super) enum DocumentOrFuture<'a> {
+    Document(Cow<'a, str>),
+    Future(DocumentFuture<'a>),
+}
+
+fn wrap_document(doc: &str) -> DocumentOrFuture<'_> {
```

Comment:
Could have been a `From`

Reply:
I don't think that would have been better